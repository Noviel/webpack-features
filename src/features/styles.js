import ExtractTextPlugin from 'extract-text-webpack-plugin';

const createRule = (
  { production, target },
  { preprocessor, cssModules, extract, postcss, exclude }
) => {
  const options = {};

  if (exclude) {
    options.exclude = exclude;
  }

  /* eslint-disable no-useless-escape */
  const extension =
    preprocessor === 'scss'
      ? `\.scss`
      : preprocessor === 'less' ? `\.less` : `\.css$`;

  const excludeModules = `^(?!.*\.module).*`;

  const test = cssModules
    ? new RegExp(`.module${extension}$`, 'i')
    : new RegExp(`${excludeModules}${extension}$`, 'i');
  /* eslint-enable no-useless-escape */

  const cssLoader = {
    loader: `css-loader${target.name === 'node' ? '/locals' : ''}`,
    options: {
      sourceMap: !production,
      minimize: production,
      modules: cssModules,
      localIdentName: production
        ? '[hash:base64]'
        : '[path][name]__[local]--[hash:base64:5]',
    },
  };

  const styleLoader = {
    loader: 'style-loader',
    options: {},
  };

  let loaders = [cssLoader];

  if (postcss) {
    loaders[0].options.importLoaders = 1;
    loaders = loaders.concat({ loader: 'postcss-loader', options: postcss });
  }

  if (preprocessor === 'scss') {
    loaders[0].options.importLoaders++;
    loaders = loaders.concat({
      loader: 'sass-loader',
      options: {
        sourceMap: !production,
      },
    });
  } else if (preprocessor === 'less') {
    loaders[0].options.importLoaders++;
    loaders = loaders.concat({
      loader: 'less-loader',
      options: {
        sourceMap: !production,
      },
    });
  }

  if (extract) {
    loaders = ExtractTextPlugin.extract({
      fallback: 'style-loader',
      use: loaders,
    });
  } else if (target.name === 'browsers') {
    loaders.unshift(styleLoader);
  }

  return {
    test,
    use: loaders,
    ...options,
  };
};

export default (
  env,
  {
    preprocessors = ['css'],
    cssModules = 'both',
    extract = env.target.name === 'browsers' && env.production,
    extractPlugin = extract,
    extractFilename = '[name].[contenthash].css',
    postcss = require('../lib/postcss.config.js'),
    exclude = false,
  },
  { plugins, next }
) => {
  const rules = [];
  const webpackPlugins = [];

  if (postcss) {
    postcss = postcss(env);
  }

  const useCSSModules = cssModules === 'both' || cssModules === 'only';
  const useGlobalCSS =
    cssModules === 'both' || cssModules === 'exclude' || true;

  preprocessors.forEach(preprocessor => {
    const options = {
      preprocessor,
      extract,
      postcss,
      exclude,
    };

    if (useCSSModules) {
      options.cssModules = true;
      rules.push(createRule(env, options));
    }
    if (useGlobalCSS) {
      options.cssModules = false;
      rules.push(createRule(env, options));
    }
  });

  if (extractPlugin) {
    webpackPlugins.push(
      new ExtractTextPlugin({
        filename: extractFilename,
      })
    );
  }

  return next(env, plugins, {
    module: {
      rules,
    },
    plugins: webpackPlugins,
  });
};
