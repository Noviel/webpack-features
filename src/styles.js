import ExtractTextPlugin from 'extract-text-webpack-plugin';

const createRule = (
  { production, target },
  { preprocessor, cssModules, extract, postcss }
) => {
  const options = {
    exclude: [/node_modules/],
  };

  /* eslint-disable no-useless-escape */
  const extension =
    preprocessor === 'scss'
      ? `\.scss`
      : preprocessor === 'less' ? `\.less` : `\.css$`;

  const excludeModules = `^(?!.*\.module).*`;

  const test = cssModules
    ? new RegExp(`.module${extension}$`, 'i')
    : new RegExp(`${excludeModules}${extension}$`, 'i');
  /* eslint-enable no-useless-escape */

  const cssLoader = {
    loader: `css-loader${target.node ? '/locals' : ''}`,
    options: {
      sourceMap: !production,
      minimize: production,
      modules: cssModules,
      localIdentName: production
        ? '[hash:base64]'
        : '[path][name]__[local]--[hash:base64:5]',
    },
  };

  const styleLoader = {
    loader: 'style-loader',
    options: {},
  };

  let loaders = [cssLoader];

  if (postcss) {
    loaders[0].options.importLoaders = 1;
    loaders = loaders.concat({ loader: 'postcss-loader', options: postcss });
  }

  if (preprocessor === 'scss') {
    loaders[0].options.importLoaders++;
    loaders = loaders.concat({
      loader: 'sass-loader',
      options: {
        sourceMap: !production,
      },
    });
  } else if (preprocessor === 'less') {
    loaders[0].options.importLoaders++;
    loaders = loaders.concat({
      loader: 'less-loader',
      options: {
        sourceMap: !production,
      },
    });
  }

  if (extract) {
    loaders = ExtractTextPlugin.extract({
      fallback: 'style-loader',
      use: loaders,
    });
  } else if (target.browsers) {
    loaders.unshift(styleLoader);
  }

  return {
    test,
    use: loaders,
    ...options,
  };
};

export default (
  { target, production },
  {
    preprocessors = ['css'],
    cssModules = 'both',
    extract = target.browsers && production,
    extractPlugin = extract,
    extractFilename = '[name].[contenthash].css',
    postcss = require('./postcss.config.js'),
  } = {},
  state
) => {
  const rules = [];
  const plugins = [];

  if (postcss) {
    postcss = postcss({ target, production });
  }

  const useCSSModules = cssModules === 'both' || cssModules === 'only';
  const useGlobalCSS =
    cssModules === 'both' || cssModules === 'exclude' || true;

  preprocessors.forEach(preprocessor => {
    const options = {
      preprocessor,
      extract,
      postcss,
    };

    if (useCSSModules) {
      options.cssModules = true;
      rules.push(createRule({ target, production }, options));
    }
    if (useGlobalCSS) {
      options.cssModules = false;
      rules.push(createRule({ target, production }, options));
    }
  });

  if (extractPlugin) {
    plugins.push(
      new ExtractTextPlugin({
        filename: extractFilename,
      })
    );
  }

  state.addRules(rules);
  state.addPlugins(plugins);
};
