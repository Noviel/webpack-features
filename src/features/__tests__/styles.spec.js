import styles, { createTestRegExp } from '../styles';
import { callFeature, envs, captionForEnv } from '../../fixtures';

describe(`createTestRegExp`, () => {
  describe(`cssModules is 'false'`, () => {
    it(`should match filenames without '.module' before an extension`, () => {
      let test = createTestRegExp('css', false);

      expect(test.test('index.css')).toBe(true);
      expect(test.test('some-filename.hash420.css')).toBe(true);

      test = createTestRegExp('scss', false);
      expect(test.test('index.scss')).toBe(true);
      expect(test.test('some-filename.hash420.scss')).toBe(true);

      test = createTestRegExp('less', false);
      expect(test.test('index.less')).toBe(true);
      expect(test.test('some-filename.hash420.less')).toBe(true);
    });

    it(`should not match filenames with '.module' before an extension`, () => {
      let test = createTestRegExp('css', false);

      expect(test.test('index.module.css')).toBe(false);
      expect(test.test('some-filename.hash420.module.css')).toBe(false);

      test = createTestRegExp('scss', false);
      expect(test.test('index.module.scss')).toBe(false);
      expect(test.test('some-filename.hash420.module.scss')).toBe(false);

      test = createTestRegExp('less', false);
      expect(test.test('index.module.less')).toBe(false);
      expect(test.test('some-filename.hash420.module.less')).toBe(false);
    });
  });

  describe(`cssModules is 'true'`, () => {
    it(`should not match filenames without '.module' before an extension`, () => {
      let test = createTestRegExp('css', true);

      expect(test.test('index.css')).toBe(false);
      expect(test.test('some-filename.hash420.css')).toBe(false);

      test = createTestRegExp('scss', true);
      expect(test.test('index.scss')).toBe(false);
      expect(test.test('some-filename.hash420.scss')).toBe(false);

      test = createTestRegExp('less', true);
      expect(test.test('index.less')).toBe(false);
      expect(test.test('some-filename.hash420.less')).toBe(false);
    });

    it(`should match filenames with '.module' before an extension`, () => {
      let test = createTestRegExp('css', true);

      expect(test.test('index.module.css')).toBe(true);
      expect(test.test('some-filename.hash420.module.css')).toBe(true);

      test = createTestRegExp('scss', true);
      expect(test.test('index.module.scss')).toBe(true);
      expect(test.test('some-filename.hash420.module.scss')).toBe(true);

      test = createTestRegExp('less', true);
      expect(test.test('index.module.less')).toBe(true);
      expect(test.test('some-filename.hash420.module.less')).toBe(true);
    });
  });
});

describe(`styles`, () => {
  for (const env in envs) {
    it(`should return a valid rule with default paremeters ${captionForEnv(
      envs[env]
    )}`, () => {
      expect(callFeature(styles, {}, [], envs[env])).toMatchSnapshot();
    });
  }
});
