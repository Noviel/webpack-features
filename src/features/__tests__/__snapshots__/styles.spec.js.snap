// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styles should return a valid rule with default paremeters for current node in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\.module\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "css-loader/locals",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "css-loader/locals",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": false,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`styles should return a valid rule with default paremeters for legacy browsers in development mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\.module\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "style-loader",
            "options": Object {},
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "style-loader",
            "options": Object {},
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": false,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`styles should return a valid rule with default paremeters for legacy browsers in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\.module\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "C:\\\\projects\\\\dev\\\\webpack-features\\\\node_modules\\\\extract-text-webpack-plugin\\\\dist\\\\loader.js",
            "options": Object {
              "omit": 1,
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "C:\\\\projects\\\\dev\\\\webpack-features\\\\node_modules\\\\extract-text-webpack-plugin\\\\dist\\\\loader.js",
            "options": Object {
              "omit": 1,
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": false,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ExtractTextPlugin {
      "filename": "[name].[contenthash].css",
      "id": 2,
      "options": Object {},
    },
  ],
}
`;

exports[`styles should return a valid rule with default paremeters for modern browsers in development mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\.module\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "style-loader",
            "options": Object {},
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": true,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "style-loader",
            "options": Object {},
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[path][name]__[local]--[hash:base64:5]",
              "minimize": false,
              "modules": false,
              "sourceMap": true,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [],
}
`;

exports[`styles should return a valid rule with default paremeters for modern browsers in production mode 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\.module\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "C:\\\\projects\\\\dev\\\\webpack-features\\\\node_modules\\\\extract-text-webpack-plugin\\\\dist\\\\loader.js",
            "options": Object {
              "omit": 1,
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": true,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "test": /\\.css\\$/i,
        "use": Array [
          Object {
            "loader": "C:\\\\projects\\\\dev\\\\webpack-features\\\\node_modules\\\\extract-text-webpack-plugin\\\\dist\\\\loader.js",
            "options": Object {
              "omit": 1,
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 1,
              "localIdentName": "[hash:base64]",
              "minimize": true,
              "modules": false,
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "plugins": Array [
                [Function],
                [Function],
              ],
              "sourceMap": false,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ExtractTextPlugin {
      "filename": "[name].[contenthash].css",
      "id": 1,
      "options": Object {},
    },
  ],
}
`;
